### PR TITLE
fix(advance-filtering): handle error when columns are dynamically changed - 18.1.x

### LINKED ISSUE #14314: Grid Advanced Filtering throws if grid has recreated its columns colection
## Steps to reproduce  

1. Add grid with Advanced filtering (you can use the `GridFilteringComponent` in dev demos
2. Through Advanced filtering dialog, filter some column, lets say `CompanyName` 
3. Change the collection, which the grid uses to create columns based on;

this.columns = this.columns2; // `columns2` has different columns than `columns`

OR change the grid data and have tje grid auto generate its columns:

```
        this.grid1.shouldGenerate = true;
        this.data = this.data3;
```

4. Open Advanced Filtering dialog

## Result  
Error is thrown and Advanced filtering dialog is not opened
image

### PR DESCRIPTION
**Pull request by georgianastasov**
Closes #14314  

### Additional information (check all that apply):
 - [x] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/grid/grid-filtering-advanced.spec.ts [hunk 1]
```diff
@@ -2048,6 +2051,84 @@ describe('IgxGrid - Advanced Filtering #grid - ', () => {
...
         }));
 
+        it('should handle advanced filtering correctly when grid columns and data are dynamically changed', fakeAsync(() => {
+            const fixture = TestBed.createComponent(IgxGridAdvancedFilteringDynamicColumnsComponent);
+            grid = fixture.componentInstance.grid;
+            fixture.detectChanges();
+
+            grid.height = '800px';
+            fixture.detectChanges();
+            tick(50);
```
Comment (ddincheva) [inline projects/igniteui-angular/src/lib/grids/grid/grid-filtering-advanced.spec.ts:2061]: if this is the desired grid height set it initially .. the grid resizing is not part of the main tested scenario ..

File: projects/igniteui-angular/src/lib/grids/grid/grid-filtering-advanced.spec.ts [hunk 2]
```diff
@@ -2048,6 +2051,84 @@ describe('IgxGrid - Advanced Filtering #grid - ', () => {
...
+                { field: 'ContactName', header: 'Contact Name', width: '200px', type: 'string' },
+                { field: 'ContactTitle', header: 'Contact Title', width: '200px', type: 'string' },
+                { field: 'City', header: 'City', width: '200px', type: 'string' },
+                { field: 'Country', header: 'Country', width: '200px', type: 'string' },
+            ];
+            fixture.detectChanges();
+            flush();
+
+            // Change the grid's data collection
+            setTimeout(() => {
```
Comment (ddincheva) [inline projects/igniteui-angular/src/lib/grids/grid/grid-filtering-advanced.spec.ts:2110]: please avoid using setTimeout

**Review by ddincheva — APPROVED** [review]

**Comment by ddincheva** [comment]
Since the PR to master with these changes is closed, I am leaving links to the comments and the remaining concerns here:
https://github.com/IgniteUI/igniteui-angular/pull/14525#issuecomment-2242679103
https://github.com/IgniteUI/igniteui-angular/pull/14525#issuecomment-2245234690
Please log an issue with the new task regarding the UX experience when filtering is applied and the data of the grid is changed and link it here .. in order to have better tracking

**Comment by georgianastasov** [comment]
I have logged the new task regarding the UX experience when filtering is applied and the data of the grid is changed. You can track it here: [#14595](https://github.com/IgniteUI/igniteui-angular/issues/14595).

**Pull request merged**